x/delegation/keeper/share.go

```diff
@@ -0,0 +1,233 @@
+package keeper
+
+import (
+	errorsmod "cosmossdk.io/errors"
+	sdkmath "cosmossdk.io/math"
+	assetstype "github.com/ExocoreNetwork/exocore/x/assets/types"
+	delegationtypes "github.com/ExocoreNetwork/exocore/x/delegation/types"
+	sdk "github.com/cosmos/cosmos-sdk/types"
+)
+
+// TokensFromShares calculate the token amount of provided shares, then truncated to Int
+// It uses `LegacyDec.Quo` to calculate the quotient, `LegacyDec.Quo` perform a bankers
+// rounding quotient, so the calculated token amount may be either larger or smaller
+// due to precision rounding issues. But it's acceptable, because bankers rounding balances the
+// deviation caused by precision, especially when a large number of restakers undertake
+// undelegation. Additionally, the last undelegation from an operator will undelegate all
+// remaining token to avoid the calculated token amount is bigger than the remaining token
+// caused by the bankers rounding.
+func TokensFromShares(stakerShare, totalShare sdkmath.LegacyDec, totalAmount sdkmath.Int) (sdkmath.Int, error) {
+	if stakerShare.GT(totalShare) {
+		return sdkmath.NewInt(0), errorsmod.Wrapf(delegationtypes.ErrInsufficientShares, "the stakerShare is:%v the totalShare is:%v", stakerShare, totalShare)
+	}
+	if totalShare.IsZero() {
+		return sdkmath.NewInt(0), delegationtypes.ErrDivisorIsZero
+	}
+	return (stakerShare.MulInt(totalAmount)).Quo(totalShare).TruncateInt(), nil
+}
+
+// SharesFromTokens returns the shares of a delegation given a delegated amount. It
+// returns an error if the validator has no tokens.
+// It uses `LegacyDec.QuoInt` to calculate the quotient, the calculated result will
+// be truncated through the truncation implemented by golang's standard big.Int.
+// So the calculated share might tend to be smaller, but it seems acceptable, because
+// we need to make sure the staker can't get a bigger share than they should get.
+func SharesFromTokens(totalShare sdkmath.LegacyDec, stakerAmount, totalAmount sdkmath.Int) (sdkmath.LegacyDec, error) {
+	if stakerAmount.GT(totalAmount) {
```

Comment:
Why `stakerAmount` greater than `totalAmount` result in an error return ?

Reply:
Great catch, this check should be removed. I will fix it in the next commit